### PR TITLE
[94x] L1T TwinMux

### PR DESCRIPTION
**Pull request by rekovic**
Emulation of TwinMux needed for 94x MC.

It moves the code from L1Trigger/LTMuonBarrel to new package L1Trigger/L1TTwinMux and uses the latest algorithm.

### CODE REVIEW

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by arunhep** [comment]
@rekovic i think this should be tested with #20758 otherwise the whole effort of updating the GTs with RPCLink maps will go in vain.

**Comment by rekovic** [comment]
@arunhep You are right, can already take advantage of the GT.  Thanks.

File: L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py [hunk 1]
```diff
@@ -0,0 +1,29 @@
+import FWCore.ParameterSet.Config as cms
+
+RPCLinkMapSource = cms.ESSource('PoolDBESSource'
+                                , DBParameters = cms.PSet()
+                                , timetype = cms.string('timestamp')
+                                , toGet = cms.VPSet(
+                                    cms.PSet(
+                                        record = cms.string('RPCLBLinkMapRcd')
+                                        , tag = cms.string('RPCLBLinkMap_L1T_v1')
```
Comment (arunhep) [inline L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py]: All these should be changed. These are part of GTs in PR #20758

**Comment by rekovic** [comment]
please test with #20758

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
Layer1 packer and unpacker will come in a separate PR.

Now test just TwinMux.

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
last commit fixes the SegFaults

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
please test

**Comment by rekovic** [comment]
+1

File: L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py (hunk 1 above)
Comment (arunhep) [reply L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py]: @rekovic why do you need to add the tags here when you have them all in GTs?

File: L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+                                    )
+                                    , cms.PSet(
+                                        record = cms.string('RPCDCCLinkMapRcd')
+                                        , tag = cms.string('RPCDCCLinkMap_L1T_v1')
+                                    )
```
Comment (arunhep) [inline L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py]: here as well

File: L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+                                    , cms.PSet(
+                                        record = cms.string('RPCTwinMuxLinkMapRcd')
+                                        , tag = cms.string('RPCTwinMuxLinkMap_L1T_v1')
+                                    )
```
Comment (arunhep) [inline L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py]: and here

File: L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py [hunk 4]
```diff
@@ -0,0 +1,29 @@
...
+                                    , cms.PSet(
+                                        record = cms.string('RPCCPPFLinkMapRcd')
+                                        , tag = cms.string('RPCCPPFLinkMap_L1T_v1')
+                                    )
```
Comment (arunhep) [inline L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py]: here

File: L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py [hunk 5]
```diff
@@ -0,0 +1,29 @@
...
+                                    , cms.PSet(
+                                        record = cms.string('RPCOMTFLinkMapRcd')
+                                        , tag = cms.string('RPCOMTFLinkMap_L1T_v1')
+                                    )
```
Comment (arunhep) [inline L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py]: and here

Comment (rekovic) [reply L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py]: Hi @arunhep 
The file L1Trigger/L1TTwinMux/python/RPCLinkMap_CondDB_cff.py is not used and its modules are not imported anywhere in this PR, so having this config file does not hurt anything.  
All the RPC*LinkMaps are accessed from GTs, as it should be.

**Comment by rekovic** [comment]
Hi @arunhep please sign or comment.

**Comment by arunhep** [comment]
+1
with following comments : 
- while checking the comparisons, I see some cms code rules violations : https://cmssdt.cern.ch/SDT/jenkins-artifacts/pull-request-integration/PR-20832/23797/codeRules/cmsCodeRule1.txt
- Other differences seen in the comparisons are because of other commits which were tested along with this PR.

File: L1Trigger/L1TTwinMux/interface/AlignTrackSegments.h [hunk 6]
```diff
@@ -0,0 +1,48 @@
...
+//
+//   AlignTrackSegments
+//
+//
+//   Author :
+//   G. Flouris               U Ioannina    Mar. 2015
+//   mod.: G Karathanasis
+//--------------------------------------------------
+
+#ifndef AlignTrackSegments_H
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/interface/AlignTrackSegments.h]: hi @rekovic - add the subsystem and package to the include guards

File: L1Trigger/L1TTwinMux/interface/AlignTrackSegments.h [hunk 7]
```diff
@@ -0,0 +1,48 @@
...
+
+class AlignTrackSegments  {
+public:
+  AlignTrackSegments(L1MuDTChambPhContainer inphiDigis);
+  ~AlignTrackSegments() {};
+
+  void run(const edm::EventSetup& c);
+
+ ///Return Output PhContainer
+ L1MuDTChambPhContainer getDTContainer(){  return m_dt_tsshifted;}
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/interface/AlignTrackSegments.h]: consider providing a const & interface

File: L1Trigger/L1TTwinMux/interface/IOPrinter.h [hunk 8]
```diff
@@ -0,0 +1,41 @@
...
+#include "Geometry/RPCGeometry/interface/RPCRoll.h"
+#include "Geometry/RPCGeometry/interface/RPCGeometry.h"
+#include "CondFormats/DataRecord/interface/RPCEMapRcd.h"
+#include "Geometry/Records/interface/MuonGeometryRecord.h"
+#include "L1Trigger/L1TTwinMux/interface/AlignTrackSegments.h"
+#include "L1Trigger/L1TTwinMux/interface/RPCtoDTTranslator.h"
+#include "L1Trigger/L1TTwinMux/interface/DTRPCBxCorrection.h"
+#include "L1Trigger/L1TTwinMux/interface/RPCHitCleaner.h"
+
+using namespace std;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/interface/IOPrinter.h]: remove

File: L1Trigger/L1TTwinMux/interface/IOPrinter.h [hunk 9]
```diff
@@ -0,0 +1,41 @@
...
+
+class IOPrinter{
+public:
+  IOPrinter() {};
+  ~IOPrinter() {};
+  void run(edm::Handle<L1MuDTChambPhContainer>, L1MuDTChambPhContainer ,  edm::Handle<RPCDigiCollection>,
+            const edm::EventSetup& );
+  void run(L1MuDTChambPhContainer*, L1MuDTChambPhContainer , RPCDigiCollection*,
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/interface/IOPrinter.h]: L1MuDTChambPhContainer should be passed by const &

File: L1Trigger/L1TTwinMux/interface/IOPrinter.h [hunk 10]
```diff
@@ -0,0 +1,41 @@
...
+#include "L1Trigger/L1TTwinMux/interface/DTRPCBxCorrection.h"
+#include "L1Trigger/L1TTwinMux/interface/RPCHitCleaner.h"
+
+using namespace std;
+
+class IOPrinter{
+public:
+  IOPrinter() {};
+  ~IOPrinter() {};
+  void run(edm::Handle<L1MuDTChambPhContainer>, L1MuDTChambPhContainer ,  edm::Handle<RPCDigiCollection>,
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/interface/IOPrinter.h]: pass things by const reference

File: L1Trigger/L1TTwinMux/interface/RPCHitCleaner.h [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+
+class RPCHitCleaner  {
+public:
+  RPCHitCleaner(RPCDigiCollection inrpcDigis);
+  ~RPCHitCleaner() {};
+
+  void run(const edm::EventSetup& c);
+
+ ///Return Output RPCCollection
+ RPCDigiCollection getRPCCollection(){  return m_outrpcDigis;}
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/interface/RPCHitCleaner.h]: use consistent indentation

File: L1Trigger/L1TTwinMux/plugins/BuildFile.xml [hunk 12]
```diff
@@ -0,0 +1,9 @@
+<library   file="*.cc" name="L1TriggerL1TTwinMuxPlugins">
+<use name="FWCore/ServiceRegistry"/>
+<use name="CondFormats/DataRecord"/>
+<use name="CondFormats/L1TObjects"/>
+<use name="L1Trigger/L1TTwinMux"/>
+<flags LDFLAGS="-fPIC"/>
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/plugins/BuildFile.xml]: remove

File: L1Trigger/L1TTwinMux/src/AlignTrackSegments.cc [hunk 13]
```diff
@@ -0,0 +1,67 @@
...
+
+
+void AlignTrackSegments::run( const edm::EventSetup& c) {
+
+
+
+  std::vector<L1MuDTChambPhDigi> l1ttma_out;
+
+ L1MuDTChambPhDigi const* ts1=nullptr;
+ L1MuDTChambPhDigi const* ts2=nullptr;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/AlignTrackSegments.cc]: just declare these within the loop (line 43)

File: L1Trigger/L1TTwinMux/src/AlignTrackSegments.cc [hunk 14]
```diff
@@ -0,0 +1,67 @@
...
+//  std::auto_ptr<const L1MuDTChambPhDigi> ts1(new L1MuDTChambPhDigi);
+ // std::unique_ptr<L1MuDTChambPhDigi> ts2(new L1MuDTChambPhDigi);
+
+  int bx=0, wheel=0, sector=0, station=1;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/AlignTrackSegments.cc]: declare these in the loops

File: L1Trigger/L1TTwinMux/src/AlignTrackSegments.cc [hunk 15]
```diff
@@ -0,0 +1,67 @@
...
+   for (wheel=-3;wheel<=3; wheel++ ){
+    for (sector=0;sector<12; sector++ ){
+     for (station=1; station<=4; station++){
+        ts1=nullptr; ts2=nullptr;
+        ///chPhiSegm1 reads the first ts from bx
+        ts1=m_phiDigis.chPhiSegm1(wheel,station,sector,bx);
+        ///chPhiSegm2 reads the second ts from bx-1
+        ts2 = m_phiDigis.chPhiSegm2(wheel,station,sector,bx);
+        ///Code = 7 NULL primitive
+        if(ts1 && ts1->code()!=7) {
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/AlignTrackSegments.cc]: ts1!=nullptr

File: L1Trigger/L1TTwinMux/src/DTLowQMatching.cc [hunk 16]
```diff
@@ -0,0 +1,86 @@
...
+  phiChambVectorDT=m_phiDTDigis->getContainer();
+  m_phiDTDigis_tm->setContainer(*phiChambVectorDT );
+
+  int bx=0, wheel=0, sector=0, station=1;
+   //cout<<"LowQ Matching  "<<track_seg<<endl;
+  for (wheel=-2;wheel<=2; wheel++ ){
+   for (sector=0;sector<12; sector++ ){
+    for (station=1; station<=4; station++){
+     for(bx=-3; bx<=3; bx++){
+         vector<int> delta_m, delta_p, delta_0;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/DTLowQMatching.cc]: are these used?

File: L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc [hunk 17]
```diff
@@ -0,0 +1,211 @@
...
+
+  BxCorrection(0);
+  BxCorrection(1);
+
+  m_dt_tsshifted.setContainer(m_l1ttma_out);
+}
+
+void DTRPCBxCorrection::BxCorrection(int track_seg){
+
+  L1MuDTChambPhDigi * dtts=nullptr;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: declare these where they are needed

File: L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc [hunk 18]
```diff
@@ -0,0 +1,211 @@
...
+  const std::vector<L1MuDTChambPhDigi> *phiChambVectorDT;
+  phiChambVectorDT= m_phiDTDigis.getContainer();
+  m_phiDTDigis_tm->setContainer(*phiChambVectorDT);
+  auto m_phiRPCDigis_tm=std::make_shared<L1MuTMChambPhContainer>();
+  //std::shared_ptr<L1MuTMChambPhContainer> m_phiRPCDigis_tm (new L1MuTMChambPhContainer);
+  const std::vector<L1MuDTChambPhDigi> *phiChambVectorRPC;
+  phiChambVectorRPC= m_phiRPCDigis.getContainer();
+  m_phiRPCDigis_tm->setContainer(*phiChambVectorRPC);
+
+  int bx=0, wheel=0, sector=0, station=1;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: declare in the loops

File: L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc [hunk 19]
```diff
@@ -0,0 +1,211 @@
...
+  int ibx_dtm = 0, fbx_dtm = 0;
+  int ibx_dtp = 0, fbx_dtp = 0;
+
+  for (wheel=-2;wheel<=2; wheel++ ){
+   for (sector=0;sector<12; sector++ ){
+     for (station=1; station<=4; station++){
+       bool shifted[7] = {false, false, false, false,false, false, false};
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: looks like these should be declared as const and outside the loop

File: L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc [hunk 20]
```diff
@@ -0,0 +1,211 @@
...
+}
+
+int DTRPCBxCorrection::deltaPhi(int dt_phi, int rpc2dt_phi ){
+  int delta_phi = abs( dt_phi - rpc2dt_phi );
+  return delta_phi;
+}
+
+std::vector<int> DTRPCBxCorrection::concat_delta(vector<int> delta_0, vector<int> delta_p, vector<int> delta_m){
+    vector<int> delta;
+    delta.insert(delta.end(), delta_0.begin(), delta_0.end());
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: delta.push_back?
(after reserving 3)

Comment (davidlange6) [reply L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: sorry, reserving the sum of the delta* lengths  (this is a rather expensive way to merge three vectors regardless)

File: L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc [hunk 21]
```diff
@@ -0,0 +1,211 @@
...
+    return delta;
+}
+
+int DTRPCBxCorrection::sign(float inv){
+  if(inv<0) return -1;
+  if(inv>0) return 1;
+  return 0;
+}
+
+int DTRPCBxCorrection::noRPCHits(L1MuTMChambPhContainer inCon, int bx, int wh, int sec, int st){
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: this seems to return the number of hits. nRPCHits would be the more typical function name

File: L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc [hunk 22]
```diff
@@ -0,0 +1,211 @@
...
+  const std::vector<L1MuDTChambPhDigi>* vInCon = inCon.getContainer();
+  for ( auto &i:* vInCon){
+    if  (bx == i.bxNum() && i.code() != 7 && i.whNum()==wh && i.scNum()==sec && i.stNum()==st) size++;
+  }
+
+  return size;
+}
+
+
+int DTRPCBxCorrection::flipBit(int inv){
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: move implementation to header file

File: L1Trigger/L1TTwinMux/src/L1MuTMChambPhContainer.cc [hunk 23]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+bool L1MuTMChambPhContainer::bxEmpty(int step) const {
+
+  bool empty = true;
+
+  for ( Phi_iterator i  = phiSegments.begin();
+                     i != phiSegments.end();
+                     i++ ) {
+    if  (step == i->bxNum()) empty = false;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/L1MuTMChambPhContainer.cc]: this if statement can just return  (and then you don't need the empty variable at all)

File: L1Trigger/L1TTwinMux/src/L1TTwinMuxAlgorithm.cc [hunk 24]
```diff
@@ -0,0 +1,152 @@
...
+  ///Correct(in bx) DT primitives by comparing them to RPC.
+//  DTRPCBxCorrection *rpc_dt_bx = new DTRPCBxCorrection(phiDigis,rpcHitsPhiDigis);
+  auto rpc_dt_bx=std::make_shared<DTRPCBxCorrection>(phiDigis,rpcHitsPhiDigis);
+  rpc_dt_bx->run(c);
+
+  L1MuDTChambPhContainer phiDigiscp = rpc_dt_bx->getDTContainer();
+
+  ///Add RPC primitives in case that there are no DT primitives.
+  std::vector<L1MuDTChambPhDigi> l1ttma_out;
+  L1MuDTChambPhDigi const* dtts1=nullptr;
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/L1TTwinMuxAlgorithm.cc]: usual comments about variables and loops (why repeat this complex loop structure so many times?

File: L1Trigger/L1TTwinMux/src/RPCtoDTTranslator.cc [hunk 25]
```diff
@@ -0,0 +1,323 @@
...
+    float phip1cm = p1cmPRG.phi();
+    float phim1cm = m1cmPRG.phi();
+    int direction = (phip1cm-phim1cm)/abs(phip1cm-phim1cm);
+    ///---Orientaion
+
+    return direction * roll->centreOfStrip(strip).x();
+}
+
+int RPCtoDTTranslator::bendingAngle(int xin, int xout, int phi){
+    int atanv = (int)(atan((xout-xin)/34.6) * 512);
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/RPCtoDTTranslator.cc]: what is 34.6?

File: L1Trigger/L1TTwinMux/src/RPCtoDTTranslator.cc [hunk 26]
```diff
@@ -0,0 +1,323 @@
...
+    int rvalue = atanv - phi/8;
+    return rvalue;
+}
+
+int RPCtoDTTranslator::localXX(int phi, int layer, int station){
+  //R[stat][layer] - radius of rpc station/layer from center of CMS
+  double R[2][2] = {{410.0,444.8},{492.7,527.3}};
+  double rvalue = R[station-1][layer-1]*tan(phi/4096.);
+  return rvalue;
+}
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/src/RPCtoDTTranslator.cc]: remove L1Trigger/L1TTwinMux/test/.__afsC861

File: L1Trigger/L1TTwinMux/test/runMuonBarrel.py [hunk 27]
```diff
@@ -0,0 +1,95 @@
...
+process.source = cms.Source('PoolSource',
+ fileNames = cms.untracked.vstring('file:/afs/cern.ch/work/g/gkaratha/private/bmtf/gen_samples/Singlemu_oneoverpt_100k.root')
+# ,eventsToProcess=cms.untracked.VEventRange('1:1:969-1:1:969')
+
+	                    )
+
+process.maxEvents = cms.untracked.PSet( input = cms.untracked.int32(2400))
+
+# PostLS1 geometry used
+process.load('Configuration.Geometry.GeometryExtended2015Reco_cff')
```
Comment (davidlange6) [inline L1Trigger/L1TTwinMux/test/runMuonBarrel.py]: use the DB geometry

File: L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc (hunk 19 above)
Comment (rekovic) [reply L1Trigger/L1TTwinMux/src/DTRPCBxCorrection.cc]: actually, these are variables, that change in the loop.

File: L1Trigger/L1TTwinMux/test/runMuonBarrel.py (hunk 27 above)
Comment (rekovic) [reply L1Trigger/L1TTwinMux/test/runMuonBarrel.py]: this is a test congig file for the developer.  not used anywhere.

**Comment by davidlange6** [comment]
please test

**Comment by rekovic** [comment]
please abort test

**Comment by rekovic** [comment]
please test

**Comment by arunhep** [comment]
+1

**Comment by rekovic** [comment]
+1

**Comment by rekovic** [comment]
@davidlange6 
I think this is ready.

**Comment by davidlange6** [comment]
hi @rekovic

A few residuals -otherwise its ready

- Arun asked about some GT related code that I also don't understand. Was that resolved?
- Recheck all the include guards. Many are incomplete
- concat_delta - its used one and as written it means twice as much memory allocated than just putting the three lines inline. Please just remove the function.

**Comment by rekovic** [comment]
@davidlange6 
> - Arun asked about some GT related code that I also don't understand. Was that resolved?

yes. His question were answered - calls to DB are in the config file which is not used in any workflows, it is there for developers.  Arun signed.  

> - Recheck all the include guards. Many are incomplete

Ok. Will do.

> - concat_delta - its used one and as written it means twice as much memory allocated than just putting the three lines inline. Please just remove the function.

Ok. Will do.

**Comment by davidlange6** [comment]
please test

**Comment by davidlange6** [comment]
merge

**Pull request merged**